src/recovery/mod.rs

```diff
@@ -1392,16 +1396,18 @@ where
                         // the buffer.
                         if is_closed(&epoch) {
                             if let Some(keys) = activated_epoch_to_keys_buffer.remove(&epoch) {
-                                for key in keys {
-                                    if let Some(logic) = logic_cache.remove(&key) {
-                                        let state_bytes = logic.snapshot();
-                                        // Retain logic if not a
-                                        // reset.
-                                        if let StateUpdate::Upsert(_) = state_bytes {
-                                            logic_cache.insert(key.clone(), logic);
+                                if take_snapshot {
+                                    for key in keys {
+                                        if let Some(logic) = logic_cache.remove(&key) {
```

Comment:
Thinking about this more, this will break the behavior of stateful operators. We use the return value of `logic.snapshot()` to determine if the entire logic is thrown away within `StatefulUnary`; if we don't throw it away, then state will remain even though the logic wanted it gone. Sorry I didn't catch this earlier.

This behavior requires something tricky because no method on `StatefulLogic` can cause it to "drop itself", it needs to signal one level up to `StatefulUnary` to do so.

Not sure what I was thinking before, but a better API here would maybe be to have `StatefulLogic::exec` return a third tuple item `bool` which `true` signals "throw me away". And then `StatefulLogic::snapshot` wouldn't need to deal with `StateBytes::Reset` at all. With a little more thinking, there might be something even cleaner than that. But in any case this becomes a much bigger change, though...

Reply:
Hmm, yeah, that was my concern that I mentioned yesterday around stateful operators. There's no way to peel that apart? Well, we can discuss as a team.

Reply:
I think our takeaway from meeting was "we should do a bigger change that disables all recovery code eventually, but not today and instead should look into some other performance improvements first".